usr/src/uts/armv8/rpi4/os/mbox.c

```diff
@@ -0,0 +1,203 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright 2021 Hayashi Naoyuki
+ * Copyright 2025 Michael van der Westhuizen
+ * Copyright 2025 OmniOS Community Edition (OmniOSce) Association.
+ */
+
+#include <sys/types.h>
+#include <sys/sunddi.h>
+#include <sys/sysmacros.h>
+#include <sys/stdbool.h>
+#include <sys/promif.h>
+#include <sys/ddi_impldefs.h>
+#include <sys/ddi_implfuncs.h>
+#include <sys/ddi_subrdefs.h>
+#include <sys/bcm2835_mbox.h>
+#include <sys/bcm2835_mboxreg.h>
+#include <sys/bcm2835_vcprop.h>
+#include <sys/bcm2835_vcio.h>
+
+static bool mbox_initialized = false;
+static kmutex_t mbox_lock;
+
+static ddi_dma_attr_t dma_attr = {
+	.dma_attr_version = DMA_ATTR_V0,
+	.dma_attr_addr_lo = 0x0000000000000000ull,
+	.dma_attr_addr_hi = 0x000000003FFFFFFFull,
+	.dma_attr_count_max = 0x000000003FFFFFFFull,
+	.dma_attr_align = 0x0000000000000001ull,
+	.dma_attr_burstsizes = 0x00000FFF,
+	.dma_attr_minxfer = 0x00000001,
+	.dma_attr_maxxfer = 0x000000000FFFFFFFull,
+	.dma_attr_seg = 0x000000000FFFFFFFull,
+	.dma_attr_sgllen = 1,
+	.dma_attr_granular = 0x00000001,
+	.dma_attr_flags = DDI_DMA_FLAGERR
+};
+static ddi_dma_attr_t dma_mem_attr;
+
+static caddr_t mbox_buffer;
+static paddr_t mbox_buffer_phys;
+static uintptr_t mbox_base;
+
+static int
+find_mbox(dev_info_t *dip, void *arg)
+{
+	pnode_t node = ddi_get_nodeid(dip);
+	if (node > 0) {
+		if (prom_is_compatible(node, "brcm,bcm2835-mbox")) {
+			*(dev_info_t **)arg = dip;
+			return (DDI_WALK_TERMINATE);
+		}
+	}
+	return (DDI_WALK_CONTINUE);
+}
+
+static void
+mbox_init(void)
+{
+	uint64_t base;
+	int err;
+
+	ASSERT(MUTEX_HELD(&mbox_lock));
+
+	dev_info_t *dip = NULL;
+	ddi_walk_devs(ddi_root_node(), find_mbox, &dip);
+
+	if (dip == NULL)
+		cmn_err(CE_PANIC, "mbox register is not found");
+
+	pnode_t node = ddi_get_nodeid(dip);
+	ASSERT(node > 0);
+	if (prom_get_reg_address(node, 0, &base) != 0) {
+		cmn_err(CE_PANIC,
+		    "prom_get_reg_address failed for mbox register");
+	}
+	mbox_base = SEGKPM_BASE + base;
```

Comment:
Can you use `psm_map_phys` here? That'll alloc a VA in the device arena and load it into the HAT.

I'm trying the get rid of device mappings in segkpm, since they require the loader to know how to set the mappings up.

Reply:
I've just moved the mbox stuff around but this makes sense. Will get to it next.

Reply:
Added the change as a separate commit to ease review. Once everything is set, I'll squash it with the commit that touches the mbox implementation.

Reply:
Thanks, this looks good.